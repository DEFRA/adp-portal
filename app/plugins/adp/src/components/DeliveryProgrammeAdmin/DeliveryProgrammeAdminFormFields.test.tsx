import React from 'react';
import {
  DeliveryProgrammeAdminFields,
  DeliveryProgrammeAdminFormFields,
  emptyForm,
} from './DeliveryProgrammeAdminFormFields';
import {
  FieldPath,
  FieldValues,
  UseFormReturn,
  useForm,
} from 'react-hook-form';
import { RenderResult, render, waitFor } from '@testing-library/react';
import { ErrorApi, errorApiRef } from '@backstage/core-plugin-api';
import { catalogApiRef } from '@backstage/plugin-catalog-react';
import { TestApiProvider } from '@backstage/test-utils';
import userEvent from '@testing-library/user-event';

type Context = {
  form?: UseFormReturn<DeliveryProgrammeAdminFields>;
};

async function setSelectField<
  TForm extends FieldValues,
  TPath extends FieldPath<TForm>,
>(
  result: RenderResult,
  form: UseFormReturn<TForm>,
  label: string,
  option: string,
  field: TPath,
) {
  const oldValue = form.getValues(field);
  await userEvent.click(result.getByLabelText(label));
  await waitFor(() => userEvent.click(result.getByText(option)));
  await waitFor(() => expect(form.getValues(field)).not.toBe(oldValue));
}

function Sut({
  context,
  defaultValues = emptyForm,
}: {
  readonly context: Context;
  readonly defaultValues?: DeliveryProgrammeAdminFields;
}) {
  context.form = useForm<DeliveryProgrammeAdminFields>({ defaultValues });

  return <DeliveryProgrammeAdminFormFields {...context.form} />;
}

function setup() {
  const mockErrorApi: jest.Mocked<ErrorApi> = {
    error$: jest.fn(),
    post: jest.fn(),
  };

  const mockCatalogApi = {
    getEntities: jest.fn(),
  };

  return {
    mockErrorApi,
    mockCatalogApi,
    async renderComponent(defaultValues?: DeliveryProgrammeAdminFields) {
      const context: Context = {};
      const result = render(
        <TestApiProvider
          apis={[
            [errorApiRef, mockErrorApi],
            [catalogApiRef, mockCatalogApi],
          ]}
        >
          <Sut context={context} defaultValues={defaultValues} />
        </TestApiProvider>,
      );
      await waitFor(() => expect(result.baseElement).not.toBeEmptyDOMElement());
      return { result, form: context.form! };
    },
  };
}

describe('DeliveryProgrammeAdminFormFields', () => {
  it('should render all fields correctly', async () => {
    const { renderComponent } = setup();

    const { result } = await renderComponent();

    expect(result.baseElement).toMatchSnapshot();
  });

  it('should render default fields', async () => {
    const { renderComponent } = setup();

    const fields: DeliveryProgrammeAdminFields = {
      aadEntityRefId: 'user-1234',
    };

    const { result } = await renderComponent(fields);

    expect(result.baseElement).toMatchSnapshot();
  });

  it('should write values back to the form', async () => {
    const { renderComponent, mockCatalogApi } = setup();
    mockCatalogApi.getEntities.mockResolvedValue({
      items: [
        {
          apiVersion: 'backstage.io/v1beta1',
          kind: 'User',
          metadata: {
            name: 'test-user-1',
            annotations: {
              ['graph.microsoft.com/user-id']:
                'f8699f8d-e3d5-4822-8979-6b6bce35ff16',
              ['metadata.annotations.microsoft.com/email']: 'user@email.com',
            },
          },
          spec: {
            profile: {
              displayName: 'Test User 1',
              email: 'user@email.com',
            },
          },
        },
      ],
    });

    const fields: DeliveryProgrammeAdminFields = {
      aadEntityRefId: 'f8699f8d-e3d5-4822-8979-6b6bce35ff16',
    };

    const { form, result } = await renderComponent();

    expect(result.baseElement).toMatchSnapshot('Empty');

    await setSelectField(
      result,
      form,
      'Select User',
      'Test User 1',
      'aadEntityRefId',
    );
    expect(result.baseElement).toMatchSnapshot('User selected');
    expect(form.getValues()).toMatchObject(fields);
  });
});
