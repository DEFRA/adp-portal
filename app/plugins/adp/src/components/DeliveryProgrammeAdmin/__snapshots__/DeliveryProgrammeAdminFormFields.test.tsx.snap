// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliveryProgrammeAdminFormFields should render all fields correctly 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="user_catalog_name"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="user_catalog_name"
        id="user_catalog_name-label"
      >
        Select User
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="user_catalog_name-label user_catalog_name"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="user_catalog_name"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          aria-describedby="user_catalog_name-helper-text"
          aria-hidden="true"
          aria-required="true"
          class="MuiSelect-nativeInput"
          name="user_catalog_name"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              Select User
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-marginDense"
        id="user_catalog_name-helper-text"
      >
        Select a user to assign Admin permissions for this delivery programme
      </p>
    </div>
  </div>
</body>
`;

exports[`DeliveryProgrammeAdminFormFields should render default fields 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="user_catalog_name"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-filled"
        data-shrink="true"
        for="user_catalog_name"
        id="user_catalog_name-label"
      >
        Select User
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="user_catalog_name-label user_catalog_name"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="user_catalog_name"
          role="button"
          tabindex="0"
        >
          user-1234
        </div>
        <input
          aria-describedby="user_catalog_name-helper-text"
          aria-hidden="true"
          aria-required="true"
          class="MuiSelect-nativeInput"
          name="user_catalog_name"
          tabindex="-1"
          value="user-1234"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-7 PrivateNotchedOutline-legendNotched-8"
          >
            <span>
              Select User
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled MuiFormHelperText-marginDense"
        id="user_catalog_name-helper-text"
      >
        Select a user to assign Admin permissions for this delivery programme
      </p>
    </div>
  </div>
</body>
`;

exports[`DeliveryProgrammeAdminFormFields should write values back to the form: Empty 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="user_catalog_name"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="user_catalog_name"
        id="user_catalog_name-label"
      >
        Select User
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="user_catalog_name-label user_catalog_name"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="user_catalog_name"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          aria-describedby="user_catalog_name-helper-text"
          aria-hidden="true"
          aria-required="true"
          class="MuiSelect-nativeInput"
          name="user_catalog_name"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11"
          >
            <span>
              Select User
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-marginDense"
        id="user_catalog_name-helper-text"
      >
        Select a user to assign Admin permissions for this delivery programme
      </p>
    </div>
  </div>
</body>
`;

exports[`DeliveryProgrammeAdminFormFields should write values back to the form: User selected 1`] = `
<body
  style=""
>
  <div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="user_catalog_name"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-filled Mui-focused Mui-focused"
        data-shrink="true"
        for="user_catalog_name"
        id="user_catalog_name-label"
      >
        Select User
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="user_catalog_name-label user_catalog_name"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="user_catalog_name"
          role="button"
          tabindex="0"
        >
          Test User 1
        </div>
        <input
          aria-describedby="user_catalog_name-helper-text"
          aria-hidden="true"
          aria-required="true"
          class="MuiSelect-nativeInput"
          name="user_catalog_name"
          tabindex="-1"
          value="test-user-1"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11 PrivateNotchedOutline-legendNotched-12"
          >
            <span>
              Select User
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled Mui-focused MuiFormHelperText-marginDense"
        id="user_catalog_name-helper-text"
      >
        Select a user to assign Admin permissions for this delivery programme
      </p>
    </div>
  </div>
</body>
`;
